Improve auto completion for shell commands

Sumandora opened the pull request:
This makes the very first argument of shell commands (like run-shell-command or pipe) auto complete programs instead of files.

This works relatively well for the commands themselves, however for the shell prompts (so the action pipe or pipe-to) this is pretty hard to do, because there the completions functions get the full line as input. This makes it impossible to figure out what argument you are editing, since you are missing the cursor position. (For command mode, this is handled by the command mode prompt itself, so the input for completion functions is just the argument which the user is editing right now.

Thinking about it, the pipe action might be able to be replaced using a macro that effectively does "@:pipe ", since the shell actions duplicate the commands from command mode.

For now, I handle this by basically completing the first argument to be a program and the rest are just seen as one big file. The previous behavior was to just see the entire prompt as a single big file, so I think this is sane.


https://github.com/user-attachments/assets/0fa7e3bc-702d-43bd-a5a3-417675c78f00

Nikita0x commented:
are you really coding with such fonts? (damn 🤯)

Sumandora commented:
> are you really coding with such fonts? (damn 🤯)

Yeah, I need something cozy, most fonts are just cold and depressing

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -640,4 +641,48 @@ pub mod completers {
+            // Go through the entire PATH and read all files into a vec.
+            let mut vec = std::env::var("PATH")
+                .unwrap_or("".to_owned())
+                .split(":")
```
This will work on unix but not windows. You can use https://doc.rust-lang.org/std/env/fn.split_paths.html to split the path in a crossplatform way

Sumandora commented:
Fixed, but the bigger issue (and also the reason why I haven't rebased onto master so far) is that the `command_line::Signature` does not allow for varargs, I need to express a Signature in which the first argument is a program and the rest are individuals files, raw_after also doesn't help that much, because it gives me all the tokens as a single String, but since the removal of Shellwords, I can't just split the arguments myself anymore and also I'm missing the cursor once again, essentially the recent changes from Shellwords to `command_line::Signature`s downgrades the command mode commands to the shell actions. Essentially the problems I explained the in the OP, are now present (and unavoidable) for the commands as well, not just the keybound actions.

the-mikedavis commented:
Yeah this is something I was aware of while working on #12527, see https://github.com/helix-editor/helix/blob/0ee5850016650f158dbdc9033ccdde6237e8ad50/helix-term/src/commands/typed.rs#L3731-L3733

Kakoune has a better strategy for completing in prompts including command mode: the entire line and the cursor position are parameters. I think that would be a medium/large refactor to switch our completion code to that. In the meantime a completer just for shell commands would be a nice addition. For splitting in prompts you can use `helix_core::command_line::split` in place of the shell words functions.

Sumandora commented:
> Yeah this is something I was aware of while working on #12527, see https://github.com/helix-editor/helix/blob/0ee5850016650f158dbdc9033ccdde6237e8ad50/helix-term/src/commands/typed.rs#L3731-L3733
> 
> Kakoune has a better strategy for completing in prompts including command mode: the entire line and the cursor position are parameters. I think that would be a medium/large refactor to switch our completion code to that. In the meantime a completer just for shell commands would be a nice addition - a replacement for `completers::none` in `SHELL_COMPLETER`. For splitting in prompts you can use `helix_core::command_line::split` in place of the shell words functions.

I found command_line::split before, but how do I figure out the parameter that is being edited? Do I just assume the cursor is at the end? I actually went ahead and did this but I deleted my code because I didn't want to downgrade the PR even more...

Sumandora commented:
I updated onto upstream/master, basically everything works, but the code isn't that pretty, the fact that var args aren't a thing forces me to use Tokenizer, which is a bit questionable. The shell_prompts require me to somehow do everything myself, and obviously the cursor position can't be respected, regardless if you want the path completer then you can also just take it out of the PR yourself, perhaps add me as co-author to the commit. Anyways, it merges and "works" now, just isn't pretty.
